src/commodore/util.ts

```diff
@@ -72,3 +78,105 @@ export function parseFileName(
 
   return facts;
 }
+
+function isObject(item: any) {
+  return item && typeof item === 'object' && !Array.isArray(item);
+}
+
+// Adapted from https://stackoverflow.com/a/37164538
+export function mergeConfig(base: any, config: any): any {
+  let output = clone(base);
+  if (isObject(base) && isObject(config)) {
+    Object.keys(config).forEach((key) => {
+      if (isObject(config[key])) {
+        if (!(key in base)) Object.assign(output, { [key]: config[key] });
+        else output[key] = mergeConfig(base[key], config[key]);
+      } else if (Array.isArray(config[key])) {
+        if (!(key in base)) {
+          Object.assign(output, { [key]: Array.from(config[key]) });
+        } else if (Array.isArray(output[key])) {
+          output[key].push(...config[key]);
+        } else {
+        }
+      } else {
+        Object.assign(output, { [key]: config[key] });
+      }
+    });
+  }
+  return output;
+}
+
+export var globalRepos: Map<string, RepoConfig> = new Map();
+
+export function globalRepoDir(globalRepoURL: string): string {
+  let repoUrl = new URL(globalRepoURL);
+  let repoPathInfo = parse(repoUrl.pathname);
+  let repoPath = `${repoUrl.host}${repoPathInfo.dir}/${repoPathInfo.name}`;
+  return cacheDir() + `/global-repos/${repoPath}`;
+}
+
+export async function cloneGlobalRepo(config: any): Promise<RepoConfig> {
+  if (globalRepos.has(config.globalRepoURL)) {
+    return globalRepos.get(config.globalRepoURL) as RepoConfig;
+  }
+
+  logger.info(
+    {
+      globalRepoURL: config.globalRepoURL,
+      tenantId: config.tenantId,
+    },
+    'Global repo for this tenant is not initialized yet, cloning it'
+  );
+  const dir: string = globalRepoDir(config.globalRepoURL);
+  await mkdir(dir, { recursive: true });
+  // TODO(sg): support Git-https?
+  const git: any = Git(dir, simpleGitConfig());
+  await git.clone(config.globalRepoURL, '.');
+  const globalExtraConfig = await parseGlobalRepoConfig(dir);
+  const rc: RepoConfig = {
+    dir: dir,
+    extraConfig: globalExtraConfig,
+  } as RepoConfig;
+  globalRepos.set(config.globalRepoURL, rc);
+  return rc;
+}
+
+export async function parseGlobalRepoConfig(globalDir: string): Promise<any> {
+  const globalDirConfigFile: string = globalDir + '/renovate.json';
+  let globalDirConfig: any = {};
+  try {
+    const globalConfigStr: string = (
+      await readFile(globalDirConfigFile)
+    ).toString();
+    globalDirConfig = JSON.parse(globalConfigStr) as any;
+  } catch (e: any) {
+    logger.info(
+      {
+        globalDir: globalDir,
+        exception: e,
+      },
+      'Exception while reading global repo config'
+    );
+  }
+  if (
+    globalDirConfig.commodore !== undefined &&
+    globalDirConfig.commodore.extraConfig !== undefined
```

Comment:
I gnerally find it easier to reason about code that checks for `undefined` AND `null`.

```suggestion
    globalDirConfig.commodore != null &&
    globalDirConfig.commodore.extraConfig != null
```